mrp_production_operation_injection/tests/test_mrp_production_inject_operation.py

```diff
@@ -0,0 +1,256 @@
+                ("product_attribute_value_id", "=", cls.color_att_value_white.id),
+            ]
+        )
+        tmpl_attr_value_black = cls.env["product.template.attribute.value"].search(
+            [
+                ("product_tmpl_id", "=", cls.product_template_drawer.id),
+                ("product_attribute_value_id", "=", cls.color_att_value_black.id),
+            ]
+        )
+
+        cls.env["mrp.bom.line"].create(
+            {
+                "bom_id": cls.bom_drawer.id,
+                "product_id": white_color_product_tmpl.product_variant_id.id,
+                "bom_product_template_attribute_value_ids": [
+                    fields.Command.link(tmpl_attr_value_white.id)
+                ],
+            }
+        )
+
+        cls.env["mrp.routing.workcenter"].create(
+            {
+                "name": "Painting",
+                "bom_id": cls.bom_drawer.id,
+                "workcenter_id": cls.assembly_line_workcenter.id,
+                "bom_product_template_attribute_value_ids": [
+                    fields.Command.link(tmpl_attr_value_white.id)
+                ],
+            }
+        )
+
+        white_drawer = cls.product_template_drawer.product_variant_ids.filtered(
+            lambda p: p.product_template_variant_value_ids == tmpl_attr_value_white
+        )
+        black_drawer = cls.product_template_drawer.product_variant_ids.filtered(
+            lambda p: p.product_template_variant_value_ids == tmpl_attr_value_black
+        )
+
+        return white_drawer, black_drawer
+
+    @classmethod
+    def _create_manufacturing_order(cls, product, bom=None):
+        mo_form = Form(cls.env["mrp.production"])
+        mo_form.product_id = product
+        if bom is not None:
+            mo_form.bom_id = bom
+        mo = mo_form.save()
+        return mo
+
+    @classmethod
+    def _get_injector_wizard_form(cls, production):
+        action = production.action_open_workorder_injector()
+        injector_form = Form(
+            cls.env[action["res_model"]].with_context(**action.get("context", {}))
+        )
+        return injector_form
+
+    @classmethod
+    def _inject_operation(cls, production, new_operation, previous_workorder):
+        injector_form = cls._get_injector_wizard_form(production)
+        injector_form.operation_id = new_operation
+        injector_form.workorder_id = previous_workorder
+        injector_wiz = injector_form.save()
+        injector_wiz.action_add_operation()
+
+    @classmethod
+    def _get_new_workorder(cls, previously_existing_wos, existing_wos):
+        return previously_existing_wos - existing_wos
+
+    @classmethod
+    def _record_time_tracking(cls, workorder, duration, productivity):
+        workorder_form = Form(workorder)
+        with workorder_form.time_ids.new() as time_tracking_form:
+            time_tracking_form.date_end = time_tracking_form.date_start + relativedelta(
+                seconds=duration
+            )
```

Comment:
Nitpick to avoid the direct import of third module and use Odoo API:
```suggestion
            time_tracking_form.date_end = fields.Datetime.add(
                time_tracking_form.date_start, seconds=duration
            )
```

Reply:
Didn't know about this one, nice :+1: